rust-symcrypt/src/gcm.rs

```diff
@@ -0,0 +1,377 @@
+                &mut self.state,
+                auth_data.as_ptr(),
+                auth_data.len() as symcrypt_sys::SIZE_T,
+            );
+        }
+    }
+
+    pub fn encrypt_part(&mut self, data: &[u8]) -> Vec<u8> {
+        unsafe {
+            let mut dst: Vec<u8> = vec![0u8; data.len()];
+            symcrypt_sys::SymCryptGcmEncryptPart(
+                &mut self.state,
+                data.as_ptr(),
+                dst.as_mut_ptr(),
+                dst.len() as symcrypt_sys::SIZE_T,
+            );
+            dst
+        }
+    }
+
+    pub fn encrypt_final(&mut self, tag: &mut [u8]) {
+        unsafe {
+            symcrypt_sys::SymCryptGcmEncryptFinal(
+                &mut self.state,
+                tag.as_mut_ptr(),
+                tag.len() as symcrypt_sys::SIZE_T,
+            );
+        }
+    }
+
+    pub fn decrypt_part(&mut self, data: &[u8]) -> Vec<u8> {
+        let mut dst: Vec<u8> = vec![0u8; data.len()];
+        unsafe {
+            symcrypt_sys::SymCryptGcmDecryptPart(
+                &mut self.state,
+                data.as_ptr(),
+                dst.as_mut_ptr(),
+                data.len() as symcrypt_sys::SIZE_T,
+            );
+            dst
+        }
+    }
+
+    pub fn decrypt_final(&mut self, tag: &[u8]) -> Result<(), SymCryptError> {
+        unsafe {
+            match symcrypt_sys::SymCryptGcmDecryptFinal(
+                &mut self.state,
+                tag.as_ptr(),
+                tag.len() as symcrypt_sys::SIZE_T,
+            ) {
+                symcrypt_sys::SYMCRYPT_ERROR_SYMCRYPT_NO_ERROR => Ok(()),
+                err => Err(err.into()), 
+            }
+        }
+    }
+}
+
+impl Drop for GcmState {
+    fn drop(&mut self) {
+        unsafe {
+            symcrypt_sys::SymCryptWipe(
+                ptr::addr_of_mut!(self.state) as *mut c_void,
+                mem::size_of_val(&mut self.state) as symcrypt_sys::SIZE_T,
+            );
+            symcrypt_sys::SymCryptWipe(
+                ptr::addr_of_mut!(self.expanded_key) as *mut c_void,
+                mem::size_of_val(&mut self.expanded_key) as symcrypt_sys::SIZE_T,
+            );
+        }
+    }
+}
+
+pub fn validate_gcm_parameters(
+    block_cipher: symcrypt_sys::PCSYMCRYPT_BLOCKCIPHER,
+    nonce: &[u8],
+    auth_data: &[u8],
+    data: &[u8],
+    tag: &[u8],
+) -> Result<(), SymCryptError> {
+    unsafe {
+        match symcrypt_sys::SymCryptGcmValidateParameters(
+            block_cipher,
+            nonce.len() as symcrypt_sys::SIZE_T,
+            auth_data.len() as symcrypt_sys::UINT64,
+            data.len() as symcrypt_sys::SIZE_T,
+            tag.len() as symcrypt_sys::SIZE_T,
+        ) {
+            symcrypt_sys::SYMCRYPT_ERROR_SYMCRYPT_NO_ERROR => Ok(()),
+            err => Err(err.into()),
+        }
+    }
+}
+
+pub fn gcm_encrypt(
+    key: &[u8],
+    nonce: &[u8],
+    auth_data: Option<&[u8]>,
+    src: &[u8],
+    tag: &mut [u8], // This is an out parameter
```

Comment:
See above comment on `encrypt_final()`

Reply:
After thinking further, recommend changing this to tag_length parameter and returning the tag in the Result.

Reply:
done